kernel/src/sched.rs

```diff
@@ -902,37 +903,101 @@ impl Kernel {
+                // function pointer is within process accessible memory. Per
+                // TRD104:
+                //
+                // > If the passed upcall is not valid (is outside process
+                // > executable memory...), the kernel...MUST immediately return
+                // > a failure with a error code of `INVALID`.
+                let rval1 = ptr.map_or(None, |upcall_ptr_nonnull| {
+                    if !process.is_valid_upcall_function_pointer(upcall_ptr_nonnull) {
+                        Some(ErrorCode::INVAL)
+                    } else {
+                        None
+                    }
                 });
-                let rval = platform.with_driver(driver_number, |driver| match driver {
-                    Some(d) => {
-                        let res = d.subscribe(subdriver_number, upcall, process.processid());
-                        match res {
-                            // An Ok() returns the previous upcall, while
-                            // Err() returns the one that was just passed
-                            // (because the call was rejected).
-                            Ok(oldcb) => oldcb.into_subscribe_success(),
-                            Err((newcb, err)) => newcb.into_subscribe_failure(err),
+
+                // If the upcall is either null or valid, then we continue
+                // handling the upcall.
+                let rval = match rval1 {
+                    Some(err) => upcall.into_subscribe_failure(err),
+                    None => {
+                        // At this point we must save the new upcall and return the
+                        // old. The upcalls are stored by the core kernel in the
+                        // grant region so we can guarantee a correct upcall swap.
+                        // However, we do need help with initially allocating the
+                        // grant if this driver has never been used before.
+                        //
+                        // To avoid the overhead with checking for process liveness
+                        // and grant allocation, we assume the grant is initially
+                        // allocated. If it turns out it isn't we ask the capsule to
+                        // allocate the grant.
+                        match crate::grant::subscribe(process, upcall) {
+                            Ok(old_upcall) => old_upcall.into_subscribe_success(),
+                            Err((new_upcall, _err)) => {
+                                // If we get here the subscribe swap failed. We try
+                                // again after asking the capsule to allocate the
+                                // grant.
+                                platform.with_driver(driver_number, |driver| match driver {
+                                    Some(d) => {
+                                        if d.allocate_grant(process.processid()).is_err() {
```

Comment:
we should probably check if the grant is allocated before calling this

Reply:
We do on line 915.

Reply:
Do we? This is what I see for line 915:
```
                // First check if `upcall_ptr` is null. A null upcall pointer
                // represents the special "unsubscribe" operation.
                let ptr = NonNull::new(upcall_ptr);
```

Reply:
Ha that's what I get for doing it on my phone. That should be "951". Basically we first try the subscribe, and only if it fails do we try to do the grant allocation and eventually retry the subscribe.